### ISSUE TITLE
Setup replaces full %PATH% instead of adding "scripts" path on windows without local admin permissions

### ISSUE THREAD

**Issue by baklarzej** [devonfw/ide#393]
**Category: Bug**

**Priority: High / Medium (depending on whether it's common)**

devonfw-ide Version 3.3.0, 2020-03-17_17.00.14
Windows 7 Professional Version 6.1 (Build 7601: Service Pack 1)
One instance of the same behaviour on Win10 confirmed.

**Expected:** 
Setting up devonfw-ide (setup.bat) should add "scripts" path (I guess "%USERPROFILE%\scripts" is intended?) to PATH Windows user environment variable.

**Actual:**
Setup erases all of the previous contents of PATH and replaces it with ";%USERPROFILE%\scripts" (there is a leading semicolon). This causes all of my previous CLI programs to not be recognised by Windows and I need to readd everything manually. Moreover, if I uninstall and reinstall devon again, the same will happen.
![image](https://user-images.githubusercontent.com/28765790/78364778-547dfb80-75be-11ea-9dc9-0ea6dd64d959.png)

**Comment by hohwille** [comment]
OMG. Sorry to hear that. Seems something went terribly wrong here.
Can others confirm this? I did not observe this but can only test with windows in VM as I do have MacOS. Would be great if some native windows user could assist here.

**Comment by hohwille** [comment]
Logic responsible for this can be found here:
https://github.com/devonfw/ide/blob/cbaabfff9916f2d4f0b17741cff13ba5e699a8bf/scripts/src/main/resources/scripts/devon.bat#L18

**Comment by hohwille** [comment]
@baklarzej thanks for your bug report. I assume this is not a general bug that happens for every user. Do you have Admin permissions on your machine or not? Are there any other things special on your machine that could have an impact.
Ideally you could run the lines that determine the user path from the linked script on your machine to a new `test.bat` file and run it to debug what does wrong here:
```
for /F "tokens=2* delims= " %%f IN ('reg query HKCU\Environment /v PATH ^| findstr /i path') do set USER_PATH=%%g
if "%USER_PATH:~-1,1%" == ";" (
  set "USER_PATH=%USER_PATH:~0,-1%"
)
echo %USER_PATH%
```

**Comment by hohwille** [comment]
@markusschuh do you have any ideas why the reg command may fail to get the current user path?

**Comment by baklarzej** [comment]
Starting with following sample content in user PATH:
`C:\apache-maven-3.5.4\bin`

After running the test.bat I got following output:

```
C:\Users\szybal\Desktop>for /F "tokens=2* delims= " %f IN ('reg query HKCU\Environment /v PATH | findstr /i path') do set USER_PATH=%g

C:\Users\szybal\Desktop>set USER_PATH=C:\apache-maven-3.5.4\bin

C:\Users\szybal\Desktop>if "n" == ";" (set "USER_PATH=C:\apache-maven-3.5.4\bi" )

C:\Users\szybal\Desktop>echo C:\apache-maven-3.5.4\bin
C:\apache-maven-3.5.4\bin
```

I also run setup.bat as Administrator (with Windows prompt whether to run as an Administator),  I got the same result as before (PATH replaced)

**Comment by hohwille** [comment]
The output of your script looks perfectly corrent.
With the limited script no `setx` is executed and therefore PATH should only be read and not be manipulated. So can you confirm that so far this code is working correct?

> I got the same result as before (PATH was replaced)

I am not quite clear what you mean with this. Or are you saying that even the limited script did "replace" the PATH?

**Comment by baklarzej** [comment]
> I am not quite clear what you mean with this. Or are you saying that even the limited script did "replace" the PATH?

No. I meant the basic `setup.bat` script, not the limited `test.bat`. I wrote about launching it as Administrator, because you asked if I have Adminstrator rights on my computer.

`test.bat` doesn't replace anything and gives output as I wrote above.

**Comment by hohwille** [comment]
So if the `PATH` is determined properly by `test.bat` what does exactly what `setup.bat` does to determine the `PATH` what else can go wrong or how can we reproduce your bug? If we cannot reproduce it, we can not make progress and fix it.

**Comment by hohwille** [comment]
We need your feedback. Otherwise we have to close this issue as invalid...

**Comment by baklarzej** [comment]
Then it seems to depend on local administration policy/user setup. I think so because one other person that could reproduce it has a similar corporate setup as me, but it was not reproducible for you. Then the problem is not strictly with devon itself and you can close it.

**Comment by hohwille** [comment]
@baklarzej thanks for your feedback. I will try to find a Windows user who has no local admin permission and can properly test this and if reproducible do further analysis on the problem. If not reproducible we will close this issue. However, my intention is that we can also support `devonfw-ide` on windows without local admin permissions, but as I can not easily reproduce such setup and this is uncommon for most of our users we still have low priority on this.

**Comment by markusschuh** [comment]
I assume, issue could be related to the

```
setx PATH "%USER_PATH%;%%USERPROFILE%%\scripts"`
```

within `devon.bat`. The `PATH` variable in Windows is the concatenation of two registry values 

1. system PATH - `HKLM\SYSTEM\CurrentControlSet\Control\Session Manager\Environment\PATH`
1. user PATH - `HKCU\Environment\PATH`

The [setx](https://docs.microsoft.com/de-de/windows-server/administration/windows-commands/setx) per default writes to the _local environment_  that is the user-`PATH`.  But if `setx` would write to the system-`PATH` here by error - and user-`PATH` were empty - this would explain the observed result.

`setx` can write the system-`PATH`, when it runs from an elevated environment. (e.g. a CMD started with "Run as administrator") **and** when the additional `/m` option is used on `setx` command line.   (see below a test with another variable from such an elevated CMD)  But `devon.bat` does not use `/m`. So still no clear explanation, what might have happened here. 

I could imagine, that some Enterprise Software for _Endpoint Privilege Management_ could introduce a runtime environment for the scripts, which directs - after some prompt - even the `setx` writes directed to the _local registry_ into the _system registry_ 

I am unsure, if `devonfw-ide` could / should be hardened to not run into such side effects. One option might be to replace the `setx` run, with an explicit registry write to the registry location, from where the PATH value war read before (`HKCU\Environment\PATH`)  But even then there is a small risk, this registry write were redirected to the system path - when this is run inside the same non-standard windows environment.

From the idea `devonfw-ide` local administrative rights should not be necessary at all. It were possible to add some detection, if the CMD scripts run in an elevated environment. But such checks always has some risks to introduce unwanted side effects. Also IMHO the elevated rights are not the main cause here.

In Summary: A solution for this issue will need to be developed on a windows system, where this specific _Endpoint Privilege Management_  is **active**. 

--

PS: A small test, that checks, if option `/m` were necessary to redirect `setx` registry writes to system environment. Result:  _yes_

```
# reg query "HKLM\SYSTEM\CurrentControlSet\Control\Session Manager\Environment" /v MY_SETX_DEBUG_TEST
ERROR: The system was unable to find the specified registry key or value.
# setx  MY_SETX_DEBUG_TEST a_value
SUCCESS: Specified value was saved.
# reg query "HKLM\SYSTEM\CurrentControlSet\Control\Session Manager\Environment" /v MY_SETX_DEBUG_TEST
ERROR: The system was unable to find the specified registry key or value.
# setx /m  MY_SETX_DEBUG_TEST a_value
SUCCESS: Specified value was saved.
# reg query "HKLM\SYSTEM\CurrentControlSet\Control\Session Manager\Environment" /v MY_SETX_DEBUG_TEST
HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\Session Manager\Environment
    MY_SETX_DEBUG_TEST    REG_SZ    a_value
```

**Comment by hohwille** [comment]
@markusschuh thanks for the analysis and observations. I removed the blocker label, as nobody was otherwise able anymore to reproduce this issue. It seems to be a rather rare case so IMHO not a blocker. If it really happens it is very bad. I am actually planning to close this issue as we can not easily reproduce it under regular circumstances. Surely whenever someone hits the problem he is more than welcome to share any futher details and we can investigate. In case the issue is then closed, we can simply reopen.

**Comment by hohwille** [comment]
Closing as planned. If you ever happen to hit this bug please reopen and try to provide any details that could help to trace down and fix it. Unless we consider it as not reproducible.

**Comment by markusschuh** [comment]
During setup of `devonfw-ide` on a new windows host, the `setup.bat` has overwritten the user part of `PATH` instead of adding the "scripts" path at the end. Even when the environment is different - Windows 10 - this is the same bug.

Also I can see now the reason for this error: The variable USER_PATH is changed within parentheses - but since no ` DELAYEDEXPANSION` is active for those lines, the `setx` uses the value as being defined before the opening parenthesis, see https://ss64.com/nt/syntax-brackets.html.  This value is usually an empty string - what is the reason, the original value is overwritten.

**Comment by hohwille** [comment]
@markusschuh thanks for your feedback.
So you are talking about this line:
https://github.com/devonfw/ide/blob/306881afd2d924e88d18b7950988e05a39ea96b4/scripts/src/main/resources/scripts/devon.bat#L22

If your analysis is correct then we would have a general bug as we are lacking DELAYEDEXPANSION.
Can you explain why it is working for most users and only very few users are facing the error?
Is is this bug reproducible by all windows users and only 1-2 users are complaining? I asked the core team and several people confirmed that it is working on their windows machine and they can not reproduce the error.

Should we add DELAYEDEXPANSION here?
Where exactly do we need to set it? Before the FOR-loop?

**Comment by hohwille** [comment]
As the default-behaviour of CMD is sick anyhow, should we add
```
setlocal EnableDelayedExpansion
```
At the top of our `devon.bat` script and finaly have freedom with this nonsense?

**Comment by hohwille** [comment]
I finally got a windows laptop and was simply able to reproduce this bug:
* remove your `scripts` folder in your home directory
* got to environment settings and remove the `PATH` variable from your users variables
* run the `setup.bat` again

Then `USER_PATH` is empty as there is no `PATH`. However, if no user `PATH` is set, then windows defaults the `PATH` to the system path. This is where we have been mistaken as this was not obvious from the error report. The `setx` is not overriding the system path but is setting the user `PATH` to `;%USERPROFILE%\scripts`. As a result this is the entire `PATH` now and you cannot access the programs from the system `PATH` anymore. Further, as the `PATH` is synatically incorrect for windows as it starts with a semicolon, it does not even find `devon.bat` in the `scripts` folder.

As a simple fix we only need to prepend `%PATH%` to `USER_PATH` if it was empty.
It is a pitty that nobody with windows has been able to reproduce and analyse this before and it took so extra-ordinary long to get this reproducible. I will fix the bug this week and hope to ship out a new release asap.

**Comment by hohwille** [comment]
Actually, I did more testing:
The effect that the system `PATH` is not inherited only happens after the `setup.bat` was called.
When you log off and log in again, the problem goes away.
The user `PATH` then still is `;%USERPROFILE%\scripts` but the resolved `PATH` variable if you open a new CMD is then:
```
«system-path»;;C:\Users\«login»\scripts
```
I assume the first of the duplicated semicolons actually comes from the system `PATH` and then the user `PATH` starting with the semicolon is appended...

**Comment by hohwille** [comment]
> So is it correct to set the user PATH to %PATH%;%USERPROFILE%\scripts if it was previously empty?

Nope. While this works technically, it atcually duplicates the system `PATH` in your actual `PATH`.

**Comment by hohwille** [comment]
```
	if "%USER_PATH%" == "" (
      setx PATH "%%USERPROFILE%%\scripts"
	  echo "ATTENTION:"
	  echo "Your user PATH environment variable has not been previously set."
	  echo "You may need to log-off and log-in again so your PATH changes are properly applied."
	  echo "Otherwise you may get errors that devon command has not been found."
	  pause
	) else (
      setx PATH "%USER_PATH%;%%USERPROFILE%%\scripts"
	)
```

**Comment by hohwille** [comment]
#954 and this issue have turned out to be duplicates. The bug has been fixed and release 2022.11.002 is available with this fix since today.
